src/System.Reflection.Metadata/tests/Metadata/MetadataReaderTests.cs

```diff
@@ -109,6 +109,975 @@ internal static unsafe int IndexOf(byte[] peImage, byte[] toFind, int start)
+            MetadataReader reader = new MetadataReader((byte*)pinned.AddrOfPinnedObject(), peImage.Length);
+            Assert.Throws<InvalidOperationException>(() => reader.GetModuleDefinition());
+            Assert.Throws<InvalidOperationException>(() => reader.GetAssemblyDefinition());
+        }
+
+        [Fact]
+        public unsafe void GetNamespaceDefinitionRootTest()
+        {
+            byte[] peImage = (byte[])PortablePdbs.DocumentsPdb.Clone();
+            GCHandle pinned = GetPinnedPEImage(peImage);
+            MetadataReader reader = new MetadataReader((byte*)pinned.AddrOfPinnedObject(), peImage.Length);
+            Assert.True(reader.GetNamespaceDefinitionRoot().Parent.IsNil);
+        }
+
+        [Fact]
+        public unsafe void GetTypeRefTreatmentAndRowIdTest()
+        {
+            byte[] peImage = (byte[])WinRT.Lib.Clone();
+
+            GCHandle pinned = GetPinnedPEImage(peImage);
+            PEHeaders headers = new PEHeaders(new MemoryStream(peImage));
+            MetadataReader reader = new MetadataReader((byte*)pinned.AddrOfPinnedObject() + headers.MetadataStartOffset, headers.MetadataSize);
+            Assert.Equal(1, reader.GetTypeReference(TypeReferenceHandle.FromRowId(1)).ResolutionScope.RowId);
+        }
+
+        [Fact]
+        public unsafe void GetCustomAttributeTreatmentAndRowIdTest()
+        {
+            byte[] peImage = (byte[])WinRT.Lib.Clone();
+
+            GCHandle pinned = GetPinnedPEImage(peImage);
+            PEHeaders headers = new PEHeaders(new MemoryStream(peImage));
+            MetadataReader reader = new MetadataReader((byte*)pinned.AddrOfPinnedObject() + headers.MetadataStartOffset, headers.MetadataSize);
+            Assert.Equal(421, reader.GetCustomAttribute(CustomAttributeHandle.FromRowId(1)).Value.GetHeapOffset());
+        }
+
+        [Fact]
+        public unsafe void GetDeclarativeSecurityAttributeTest()
+        {
+            byte[] peImage = (byte[])WinRT.Lib.Clone();
+            GCHandle pinned = GetPinnedPEImage(peImage);
+            PEHeaders headers = new PEHeaders(new MemoryStream(peImage));
+
+            //38688266055 is the presenttable index
+            int presentTablesIndex = IndexOf(peImage, BitConverter.GetBytes(38688266055), headers.MetadataStartOffset);
+            Assert.NotEqual(presentTablesIndex, -1);
+
+            //38654728007 gives a value to the declaritive secutity row count and removes it from the row count after it.
+            Array.Copy(BitConverter.GetBytes((ulong)38654728007), 0, peImage, presentTablesIndex + headers.MetadataStartOffset, BitConverter.GetBytes((ulong)38654728007).Length);
+
+            MetadataReader reader = new MetadataReader((byte*)pinned.AddrOfPinnedObject() + headers.MetadataStartOffset, headers.MetadataSize);
+            Assert.Equal(12, reader.GetDeclarativeSecurityAttribute(DeclarativeSecurityAttributeHandle.FromRowId(1)).PermissionSet.GetHeapOffset());
+        }
```

Comment:
Tests that validate behavior on bad metadata sometimes need to tweak the bits directly since MetadataBuilder can't produce certain kinds of bad metadata. Tests that validate well-formed metadata should either use an appropriate assembly from resources or generate metadata via MetadataBuilder.

Reply:
The usage of MetadataBuilder in your last commit looks good.